x/programs/rust/wasmlanche-sdk/src/state.rs

```diff
@@ -68,30 +96,47 @@ impl State {
     /// the host fails to read the key and value.
     /// # Panics
     /// Panics if the value cannot be converted from i32 to usize.
-    pub fn get<T, K>(&self, key: K) -> Result<T, Error>
+    pub fn get<V>(&mut self, key: K) -> Result<V, Error>
     where
-        K: Into<Key>,
-        T: BorshDeserialize,
+        V: BorshDeserialize,
     {
-        let val_ptr = unsafe { host::get_bytes(&self.program, &key.into())? };
-        if val_ptr < 0 {
-            return Err(Error::Read);
-        }
+        let val_bytes = if let Some(val) = self.cache.get(&key) {
+            val.clone()
+        } else {
+            let val_ptr = unsafe { host::get_bytes(&self.program, &key.clone().into())? };
+            if val_ptr < 0 {
+                return Err(Error::Read);
+            }
+
+            // Wrap in OK for now, change from_raw_ptr to return Result
+            into_bytes(val_ptr)
+        };
 
-        // Wrap in OK for now, change from_raw_ptr to return Result
-        unsafe { from_host_ptr(val_ptr) }
+        self.cache.insert(key, val_bytes.clone());
+
+        from_slice::<V>(&val_bytes).map_err(|_| StateError::Deserialization)
     }
 
     /// Delete a value from the hosts's storage.
     /// # Errors
     /// Returns an [Error] if the key cannot be serialized
     /// or if the host fails to delete the key and the associated value
-    pub fn delete<K>(&self, key: K) -> Result<(), Error>
-    where
-        K: Into<Key>,
-    {
+    pub fn delete(&mut self, key: K) -> Result<(), Error> {
+        self.cache.remove(&key);
```

Comment:
🤔, can you add a TODO comment to cache deletes as well. We want to defer the deletes to when the `State` is dropped as well, but that requires some changes to host functionality.

Reply:
Wrote it, why would we need to change the host functionality ? Can't we modify the `get` method to check if the value was deleted and return from the cache if not, and lastly from the runtime if not present in the rw cache ?

Reply:
It would be better to pass all updates to the host at the end of the program-instance execution instead of explicitly calling `delete`. The goal is to minimize hopping back and forth when we don't need to. However, the host currently needs the functionality to handle that.